### PR TITLE
Optimize allocation of managed memory

### PR DESCRIPTION
**Pull request by AdrianSoundy**
## Description

For some reason we were getting a much smaller size of managed heap when running on IF 5.x firmware then previously even through parameters where unchanged. To try to correct this another PR was merged with a changed                  ESP32_RESERVED_RAM_FOR_IDF_ALLOCATION parameter (changed from 100 -> 50). This improved memory size but not enough.

I have investigated why this happened and adjusted code to stop it happening on future releases.

The reason is due to the IDF 5.x version having a more fragmented heap although still a similar amount of free memory.  
We were working out how much memory to allocate for managed heap based on largest free block minus amount of memory to reserve for IDF.  

This has now been changed to use the free space instead of largest free then sized to fit largest block.  
This is a more accurate value to reserve for IDF. The CmakePreset.json values have been restored to original values.

Largest free block with IDF 4.4.6 was 221184. With IDF 5.1.3 it is 114688
Free heap for 4.4.6 is 240600 and 232884 for IDF 5.1.3

## Motivation and Context
Issue reported on discord

## How Has This Been Tested?
Tested with local builds and c# test program.

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dev Containers (changes related with Dev Containers, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).

## Summary by CodeRabbit

- **New Features**
  - Enhanced memory management for ESP32 targets, ensuring better performance and stability.
  - Improved support for Bluetooth and Thread configurations.

- **Improvements**
  - Optimized memory allocation for various APIs including OneWire, Bluetooth, and I2C Slave.
  - Adjusted managed heap allocation to better utilize available memory.

- **Bug Fixes**
  - Removed outdated `ESP32_RESERVED_RAM_FOR_IDF_ALLOCATION` parameter for cleaner configuration.

### CODE REVIEW

File: targets/ESP32/_nanoCLR/Memory.cpp [hunk 1]
```diff
@@ -123,14 +123,21 @@
...
-            managedHeapSize = largestFreeBlock - INTERNAL_RAM_LEAVE_FREE_FOR_ALLOCATION;
+            // get managed heap size to allocate and if required adjust down to fit in largest block
+            managedHeapSize = freeSpaceHeap - INTERNAL_RAM_LEAVE_FREE_FOR_ALLOCATION;
+            if (managedHeapSize > largestFreeBlock)
+            {
+                managedHeapSize = largestFreeBlock;   
+            }
 
-            ESP_LOGI(TAG, "Allocating managed heap from internal RAM");
+            ESP_LOGI(TAG, "Allocating managed heap from internal RAM, total free heap %d", freeSpaceHeap);
```
[the remark this answers is not in the document]
Comment (AdrianSoundy) [reply targets/ESP32/_nanoCLR/Memory.cpp:140]: Already logged in another message

File: targets/ESP32/_nanoCLR/Memory.cpp [hunk 2]
```diff
@@ -122,20 +133,44 @@ void HeapLocation(unsigned char *&baseAddress, unsigned int &sizeInBytes)
...
+            // Reserve minimum IDF memory by subtracting off free space
+            // Everything left can be used for managed code
+            ASSERT(freeSpaceHeap > IdfToReserve);
+            managedHeapSize = (freeSpaceHeap - IdfToReserve);
+
+            // Adjust managed heap size down to fit in largest native heap block
+            if (managedHeapSize > largestFreeBlock)
+            {
+                managedHeapSize = largestFreeBlock;
+            }
```
[the remark this answers is not in the document]
Comment (AdrianSoundy) [reply targets/ESP32/_nanoCLR/Memory.cpp:167]: Not required as only used once

**Review by josesimoes — APPROVED** [review]
Looks like a nice improvement! Well done Adrian. 👍🏻

**Comment by josesimoes** [comment]
Available memory looks much better today for most targets!

**Comment by DarkNoir** [comment]
Yay, thankyou master Soundy!

**Pull request merged**